File: backend/app/models/maze.py
# =============================================================================
# models/maze.py — Pydantic schemas for Maze Runner API request/response models
# =============================================================================

# BaseModel: base class for all Pydantic models (handles validation + serialization)
from pydantic import BaseModel, Field, model_validator

# Any extra typing utilities we need
from typing import Self


# =============================================================================
# REQUEST MODELS
# These define what the API expects to receive from the frontend.
# =============================================================================

class MazeGenerateRequest(BaseModel):
    """
    Request body for POST /maze/generate.

    The frontend sends a seed (for reproducibility) and a size (grid dimensions).
    Both fields are validated before the generator ever runs.
    """

    seed: int = Field(
        ...,
        description="Random seed for reproducible maze generation. "
                    "The same seed + size always produces the same maze.",
        examples=[42],
    )

    size: int = Field(
        ...,
        description="Width and height of the square maze grid. "
                    "Must be an odd number and at least 11 "
                    "(e.g. 11, 13, 15 …). "
                    "Odd sizes ensure walls and paths align correctly.",
        ge=11,          # Pydantic v2: 'ge' = greater-than-or-equal (replaces Field(min=…))
        examples=[11, 15, 21],
    )

    # -------------------------------------------------------------------------
    # model_validator runs after all individual field validators have passed.
    # mode="after" means 'self' is already a fully constructed model instance,
    # so we can read self.size safely.
    # -------------------------------------------------------------------------
    @model_validator(mode="after")
    def size_must_be_odd(self) -> Self:
        """
        Enforce that 'size' is an odd integer.

        Maze generation algorithms (recursive backtracking, Prim's, etc.)
        require odd dimensions so that every cell sits on an odd coordinate
        and every wall sits on an even coordinate.
        """
        if self.size % 2 == 0:
            raise ValueError(
                f"'size' must be an odd number (received {self.size}). "
                "Try the next odd value: "
                f"{self.size + 1}."
            )
        return self


# =============================================================================

class MazeSolveRequest(BaseModel):
    """
    Request body for POST /maze/solve.

    The frontend sends the raw maze grid plus a start cell and goal cell.
    The pathfinder service returns the shortest path between them.
    """

    maze: list[list[int]] = Field(
        ...,
        description="2-D grid representing the maze. "
                    "0 = open path, 1 = wall.",
    )

    start: list[int] = Field(
        ...,
        description="[row, col] coordinates of the player's starting position.",
        min_length=2,   # must contain exactly two elements
        max_length=2,
        examples=[[0, 1]],
    )

    goal: list[int] = Field(
        ...,
        description="[row, col] coordinates of the target/exit cell.",
        min_length=2,
        max_length=2,
        examples=[[10, 9]],
    )


# =============================================================================
# RESPONSE MODELS
# These define exactly what the API sends back to the frontend.
# Pydantic serialises the data and strips any extra fields automatically.
# =============================================================================

class MazeGenerateResponse(BaseModel):
    """
    Response body returned by POST /maze/generate.

    Contains the full maze grid plus key coordinates so the frontend
    knows where to place the player and mark the objective.
    """

    seed: int = Field(
        ...,
        description="The seed that was used to generate this maze. "
                    "Store it on the frontend to allow 'replay same maze'.",
    )

    size: int = Field(
        ...,
        description="Width (and height) of the returned square grid.",
    )

    start: list[int] = Field(
        ...,
        description="[row, col] of the player's spawn point. "
                    "Always an open cell adjacent to the outer wall.",
    )

    center: list[int] = Field(
        ...,
        description="[row, col] of the maze's centre cell — "
                    "this is the goal / exit the player must reach.",
    )

    maze: list[list[int]] = Field(
        ...,
        description="Full 2-D grid. 0 = walkable path, 1 = wall. "
                    "Indexed as maze[row][col].",
    )


# =============================================================================

class MazeSolveResponse(BaseModel):
    """
    Response body returned by POST /maze/solve.

    Contains the optimal path from start to goal as an ordered list
    of [row, col] coordinates, plus a convenience length field.
    """

    path: list[list[int]] = Field(
        ...,
        description="Ordered list of [row, col] waypoints from start to goal "
                    "(inclusive of both endpoints). "
                    "Empty list if no path exists.",
    )

    path_length: int = Field(
        ...,
        description="Number of steps in the path (len(path) - 1). "
                    "0 if start == goal, -1 if no path was found.",
        ge=-1,
    )

File: types/maze.ts
// =============================================================================
// types/maze.ts — Single source of truth for all Maze Runner data contracts
// =============================================================================
//
// Architecture rule
// -----------------
// This file has ZERO imports and ZERO business logic.
// Every other module imports FROM here — nothing here imports from elsewhere.
//
//   types/maze.ts
//        ▲
//        │  imported by
//   ┌────┴──────────────────────────────────┐
//   │  lib/api.ts   stores/gameStore.ts     │
//   │  components/  hooks/useGame.ts        │
//   └───────────────────────────────────────┘
//
// Changing a type here is the ONLY place you need to update to propagate
// a contract change across the entire frontend.
// =============================================================================


// =============================================================================
// SECTION 1 — PRIMITIVE / REUSABLE TYPES
// These are the building blocks used by every other type in this file.
// =============================================================================

/**
 * A single [row, col] coordinate inside the maze grid.
 *
 * - row : vertical index   (0 = top row)
 * - col : horizontal index (0 = left column)
 *
 * Used for: player position, start, center/goal, and every node in a path.
 * Tuple type (not number[]) enforces exactly two elements at compile time.
 */
export type Position = [row: number, col: number];


/**
 * A single maze cell value.
 *
 *   0 → PATH  (walkable)
 *   1 → WALL  (impassable)
 *
 * Using a literal union (not plain `number`) means TypeScript will error
 * if any code tries to assign, e.g., 2 or -1 to a cell.
 */
export type CellValue = 0 | 1;


/**
 * The full maze grid — a 2-D array of CellValue.
 *
 * Constraints (enforced at runtime by the backend, documented here):
 *   - Always square: maze[i].length === maze.length for all i
 *   - Always odd dimension: 11 × 11, 21 × 21, 31 × 31, …
 *   - Outer ring is always WALL (1)
 *
 * Indexed as: maze[row][col]
 */
export type MazeGrid = CellValue[][];


/**
 * Valid odd maze dimensions accepted by the backend.
 *
 * The backend enforces size >= 11 and size % 2 !== 0.
 * Listing common sizes here makes form inputs and dropdowns type-safe;
 * extend the union as new sizes are supported.
 */
export type MazeSize = 11 | 15 | 21 | 25 | 31 | 41 | 51;


// =============================================================================
// SECTION 2 — API REQUEST TYPES
// Shapes sent FROM the frontend TO the FastAPI backend.
// Must match the Pydantic request models in app/models/maze.py exactly.
// =============================================================================

/**
 * Request body for POST /api/v1/maze/generate
 *
 * Backend Pydantic model: MazeGenerateRequest
 *   - seed : any integer — deterministic RNG seed
 *   - size : odd integer >= 11 — maze grid dimension
 */
export interface MazeGenerateRequest {
    /** Integer seed. Same seed + size always produces identical maze. */
    seed: number;

    /**
     * Width and height of the square maze grid.
     * Must be odd and >= 11 (validated by the backend).
     */
    size: MazeSize | number;   // MazeSize for known values; number for custom input
}


/**
 * Request body for POST /api/v1/maze/solve
 *
 * Backend Pydantic model: MazeSolveRequest
 *   - maze  : full grid (re-sent so the backend is stateless)
 *   - start : player's current [row, col]
 *   - goal  : target [row, col] (usually the center cell)
 */
export interface MazeSolveRequest {
    /** Full 2-D maze grid as returned by /generate. */
    maze: MazeGrid;

    /** [row, col] of the player's current position. */
    start: Position;

    /** [row, col] of the target cell (the maze center). */
    goal: Position;
}


// =============================================================================
// SECTION 3 — API RESPONSE TYPES
// Shapes received FROM the FastAPI backend BY the frontend.
// Must match the Pydantic response models in app/models/maze.py exactly.
// =============================================================================

/**
 * Response body from POST /api/v1/maze/generate
 *
 * Backend Pydantic model: MazeGenerateResponse
 */
export interface MazeGenerateResponse {
    /** Echo of the seed used — store this to allow "replay same maze". */
    seed: number;

    /** Echo of the size used — width and height of the returned grid. */
    size: number;

    /**
     * Full 2-D maze grid.
     * 0 = walkable path, 1 = wall. Indexed as maze[row][col].
     */
    maze: MazeGrid;

    /**
     * Player spawn point — always an open cell on the top inner row.
     * Place the player sprite here on game start.
     */
    start: Position;

    /**
     * Goal / exit cell — always the geometric centre of the grid.
     * e.g. size=11 → center=[5,5]. Render the objective marker here.
     */
    center: Position;
}


/**
 * Response body from POST /api/v1/maze/solve
 *
 * Backend Pydantic model: MazeSolveResponse
 */
export interface MazeSolveResponse {
    /**
     * Ordered list of [row, col] waypoints from start to goal (inclusive).
     * Produced by BFS — guaranteed to be the shortest possible path.
     * Empty array [] when no path exists.
     */
    path: Position[];

    /**
     * Number of steps in the path (path.length - 1).
     * -1 when no path exists (disconnected maze region).
     *  0 when start === goal.
     */
    path_length: number;
}


// =============================================================================
// SECTION 4 — GAME STATUS
// Union type representing every possible phase of a game session.
// Used by the Zustand store and all UI components that conditionally render
// based on the current game phase.
// =============================================================================

/**
 * All valid phases of a game session — in typical lifecycle order:
 *
 *   idle  →  loading  →  playing  →  solved
 *               ↑                       │
 *               └───────────────────────┘  (replay / new game resets to loading)
 *
 *   idle    : No maze loaded. Show the start screen / seed input.
 *   loading : Awaiting response from /generate. Show spinner.
 *   playing : Maze rendered, player can move. Timer running.
 *   solved  : Player reached the center. Show results overlay.
 *   error   : API call failed or validation error. Show error state.
 */
export type GameStatus = "idle" | "loading" | "playing" | "solved" | "error";


// =============================================================================
// SECTION 5 — FRONTEND GAME STATE
// The canonical shape of the Zustand game store.
// This type is FRONTEND-ONLY — it has no backend equivalent.
// =============================================================================

/**
 * Complete state model for the Maze Runner game session.
 *
 * Owned by: stores/gameStore.ts (Zustand)
 * Read by : components/, hooks/useGame.ts
 *
 * Field notes
 * -----------
 * - maze, start, center  : populated after a successful /generate call.
 * - player               : initialised to `start`, updated on each move.
 * - path                 : populated after a successful /solve call (hint mode).
 * - status               : drives all conditional UI rendering.
 * - seed                 : kept in state so the user can share / replay the maze.
 * - error                : set when status === "error"; null otherwise.
 * - moveCount            : incremented on every valid player move (for scoring).
 * - elapsedSeconds       : wall-clock time since the game entered "playing".
 */
export interface GameState {
    // ── Maze data (from /generate response) ──────────────────────────────
    /** Full 2-D grid. Empty array [] before first generation. */
    maze: MazeGrid;

    /** Player spawn point. Matches MazeGenerateResponse.start. */
    start: Position;

    /** Goal / exit cell. Matches MazeGenerateResponse.center. */
    center: Position;

    // ── Live game data ─────────────────────────────────────────────────────
    /** Current player position. Initialised to `start` when game begins. */
    player: Position;

    // ── Solve / hint data (from /solve response, optional) ────────────────
    /**
     * BFS shortest path from player's position to center.
     * Populated when the player requests a hint.
     * Empty array [] when no hint has been requested or no path exists.
     */
    path: Position[];

    // ── Session metadata ───────────────────────────────────────────────────
    /** Current phase of the game session. Drives all conditional UI. */
    status: GameStatus;

    /**
     * Active seed — echoed from the backend so the player can replay
     * or share an identical maze by reusing this value.
     * null before the first /generate call.
     */
    seed: number | null;

    /** Human-readable error message when status === "error". null otherwise. */
    error: string | null;

    // ── Scoring ────────────────────────────────────────────────────────────
    /** Total valid moves made since the game entered "playing". */
    moveCount: number;

    /** Seconds elapsed since the game entered "playing". */
    elapsedSeconds: number;
}


// =============================================================================
// SECTION 6 — CONVENIENCE UNIONS / UTILITY TYPES
// Derived types that other modules can import instead of re-declaring.
// =============================================================================

/**
 * The four compass directions the player can move.
 * Used by useGame.ts and any keyboard / swipe handler.
 */
export type Direction = "up" | "down" | "left" | "right";


/**
 * Minimal maze metadata — a lightweight slice of MazeGenerateResponse
 * used by components that only need seed + size (e.g. a HUD "replay" button)
 * and should not hold a reference to the full grid.
 */
export interface MazeMeta {
    seed: number;
    size: number;
}


/**
 * Payload shape for an API error response from FastAPI.
 * FastAPI's HTTPException serialises as { detail: string }.
 * Storing this type here keeps error-handling logic consistent
 * across all api.ts calls.
 */
export interface ApiErrorResponse {
    detail: string;
}